Google login passport

falsidge opened the pull request:
## Summary
Adds login to google, which allows loading and save of schedules to google account. Saving of schedule is hidden when logged in. Email is showed in popup and name and profile pic is displayed in AppBar

## Test Plan
Sign in and schedule loads with name shown.

## Issues
Closes #229 

## Future Followup (Optional)
User profile pic may not display when logged in.

falsidge committed:
Google login support

falsidge committed:
Better display that the user is signed in

falsidge committed:
Hide Save if logged in

falsidge committed:
Fix linting

ChaseC99 commented on src/components/App/AccountBase.js:
```diff
@@ -0,0 +1,37 @@
+                            alt={user.passport.user.name}
+                            src={user.passport.user.picture}
+                            style={{ marginRight: 5, width: 25, height: 25 }}
+                        ></Avatar>
```
I think this can be self closing
```suggestion
                        />
```

ChaseC99 commented on src/components/App/AccountBase.js:
```diff
@@ -0,0 +1,37 @@
+                        <Avatar
+                            alt={user.passport.user.name}
+                            src={user.passport.user.picture}
+                            style={{ marginRight: 5, width: 25, height: 25 }}
```
Let's change these styles to classNames

ChaseC99 commented on src/components/App/CustomAppBar.js:
```diff
@@ -48,7 +48,7 @@ const CustomAppBar = (props) => {
                 <div style={{ flexGrow: '1' }} />
 
-                <LoadSaveScheduleFunctionality />
+                <ScheduleLoginManager></ScheduleLoginManager>
```
Can this be self closing too?
```suggestion
                <ScheduleLoginManager />
```

ChaseC99 commented on src/components/App/AccountBase.js:
```diff
@@ -0,0 +1,37 @@
+class AccountBase extends PureComponent {
+    render() {
+        const user = this.props.user;
+        return (
```
Moving forward, everything should be functional components.

```suggestion
const AccountBase = (props) => {
    const { users } = props;
    
    return (
```

ChaseC99 commented on src/components/App/ScheduleLoginManager.js:
```diff
@@ -0,0 +1,38 @@
+class ScheduleLoginManager extends PureComponent {
+    state = {
+        user: null,
+    };
```
Convert to functional component
```suggestion
const ScheduleLoginManager = (props) => {
    const [user, setUser] = useState(null);
```

ChaseC99 reviewed (changes requested):
Since I haven't been able to test locally yet, here's some feedback on the javascript. 
This is a big PR, so great work on it so far!

ChaseC99 committed:
Merge branch 'main' into google-login-passport

ChaseC99 committed:
Convert to functional components

ChaseC99 committed:
Move Load/Log Out under Google Profile, use only one save button, rename save/loadUser to save/loadGoogleUser

ChaseC99 committed:
Remove unused imports

ChaseC99 committed:
Change save message to display email instead of passport id

ChaseC99 committed:
Fix text color in lightmode

ChaseC99 committed:
If there's an exisiting schedule loaded, add the ability to import it

falsidge committed:
Change logout to POST to avoid browser prefetch and fix localhost user icon

Voark commented on src/components/AppBar/ScheduleLoginManager.js:
```diff
@@ -1,31 +1,122 @@
+                    });
+
+                    // Add the cached data onto userData
+                    userData.addedCourses = userData.addedCourses.concat(oldData.addedCourses);
```
I don't think this works if you already have a course in your Google schedule and you try to add the same course from the old schedule. Instead we would have to add the schedule indices if the same course is in both the old and new schedules.

falsidge replied:
~~This works because of the lines above 50-70. It creates a new schedule rather than overwrite the schedules from the google login.~~

falsidge replied:
Oh I see the problem

falsidge committed:
Fix schedule merging.
Added ability to merge old schedule to google schedule.
Add console error when requests fail
Remove unnecessary json header for logout
